native_changes/RNDeviceModule.java

```diff
@@ -0,0 +1,436 @@
+            constants.put("buildNumber", preferences.getInt("buildNumber", 0));
+
+            constants.put("firstInstallTime", preferences.getLong("firstInstallTime", 0));
+            constants.put("lastUpdateTime", preferences.getLong("lastUpdateTime", 0));
+            constants.put("apiLevel", preferences.getString("apiLevel", ""));
+
+            constants.put("deviceLocale", preferences.getString("deviceLocale", ""));
+            constants.put("deviceCountry", preferences.getString("deviceCountry", ""));
+            constants.put("timezone", preferences.getString("timezone", ""));
+            constants.put("is24Hour", preferences.getBoolean("is24Hour", false));
+        }
+
+
+        return constants;
+    }
+
+    /**
+     * non-volatile data
+     * - appName
+     * - buildVersion
+     * <p>
+     * - instanceId
+     * - serialNumber
+     * - deviceName
+     * - systemName
+     * - model
+     * - brand
+     * - deviceId
+     * <p>
+     * - uniqueId
+     * - systemManufacturer
+     * - bundleId
+     * - userAgent
+     * <p>
+     * - isEmulator
+     * - isTablet
+     * <p>
+     * - phoneNumber
+     * - carrier
+     * - maxMemory
+     * - totalMemory
+     */
+    private Map<String, Object> addNonVolatileConstants(SharedPreferences preferences) {
+        Map<String, Object> constants = new HashMap<>();
+
+        constants.put("appName", preferences.getString("appName", ""));
+        constants.put("buildVersion", preferences.getString("buildVersion", ""));
+
+        constants.put("instanceId", preferences.getString("instanceId", ""));
+        constants.put("serialNumber", preferences.getString("serialNumber", ""));
+        constants.put("deviceName", preferences.getString("deviceName", ""));
+        constants.put("systemName", preferences.getString("systemName", ""));
+        constants.put("model", preferences.getString("model", ""));
+        constants.put("brand", preferences.getString("brand", ""));
+        constants.put("deviceId", preferences.getString("deviceId", ""));
+
+        constants.put("uniqueId", preferences.getString("uniqueId", ""));
+        constants.put("systemManufacturer", preferences.getString("systemManufacturer", ""));
+        constants.put("bundleId", preferences.getString("bundleId", ""));
+        constants.put("userAgent", preferences.getString("userAgent", ""));
+
+        constants.put("isEmulator", preferences.getBoolean("isEmulator", false));
+        constants.put("isTablet", preferences.getBoolean("isTablet", false));
+
+        constants.put("phoneNumber", preferences.getString("phoneNumber", ""));
+        constants.put("carrier", preferences.getString("carrier", ""));
+        constants.put("maxMemory", preferences.getLong("maxMemory", 0));
+        constants.put("totalMemory", preferences.getLong("totalMemory", 0));
+
+        return constants;
+    }
+
+    private void saveConstants(Map<String, Object> constants, Context ctx) {
```

Comment:
Saves constants into Android's key/value store. 

DeviceInfo, in particular, creates an overhead to initialize:

```
constants.put("userAgent", WebSettings.getDefaultUserAgent(this.reactContext));
```

This is a constant we know won't change in a device therefore we can cache without impacting functionality in the app